moreorless/combined.py

```diff
@@ -0,0 +1,221 @@
+        to_filenames = [f"b/{basename}"] * len(to_files)
+
+    assert len(from_files) == len(from_filenames)
+    assert len(to_files) == len(to_filenames)
+    assert len(from_files) == 1 or len(to_files) == 1
+    merge_mode = len(to_files) == 1
+
+    file_header_symbols = ["-"] * len(from_files) + ["+"] * len(to_files)
+
+    buf: List[str] = []
+    for sym, fn in zip(file_header_symbols, from_filenames + to_filenames):
+        buf.append(f"{sym}{sym}{sym} {fn}\n")
+
+    for block in _group(_contributions(from_files, to_files), context=context):
+        if isinstance(block, Hunk):
+            block.to_buf(buf, merge_mode)
+
+    return "".join(buf)
+
+
+def _contributions(
+    from_files: Sequence[str],
+    to_files: Sequence[str],
+) -> Sequence[Line]:
+    """
+    Calculate the contribution of various lines to the final result.
+
+    Instead of storing the +/- symbols directly, we will store 1 in the
+    column if the line exists in that file.
+
+    This intermediate form is very simple to look for snip points (more than
+    2*context runs of `(1, ...)`), and has the nice property that once
+    you transpose to get columns, the `@@` lines for indices `i:j` start at
+    `sum(col[:i])` and are `sum(col[i:j])` long.
+    """
+    if len(to_files) == 1:
+        common = to_files[0]
+        rest = from_files
+        merge_mode = True
+    elif len(from_files) == 1:
+        common = from_files[0]
+        rest = to_files
+        merge_mode = False
+    else:
+        raise ValueError("At least one of from_files or to_files must be length one.")
+
+    common_template = [0] * len(rest) + [1]
+
+    common_lines = common.splitlines(True)
+    common_contributions = [common_template.copy() for _ in common_lines]
+
+    # dest_idx: {text: {idx: one_or_zero}}
+    inserted_lines: List[Dict[str, Dict[int, int]]] = [
+        {} for _ in range(len(common_lines) + 1)
+    ]
+
+    def helper(s: str, dest_idx: int) -> Dict[int, int]:
+        if s not in inserted_lines[dest_idx]:
+            inserted_lines[dest_idx][s] = {}
+        return inserted_lines[dest_idx][s]
+
+    for si, src in enumerate(rest):
+        src_lines = src.splitlines(True)
+        for tag, i1, i2, j1, j2 in SequenceMatcher(
+            None, common_lines, src_lines
+        ).get_opcodes():
+            for i in range(i1, i2):
+                if tag == "equal":
+                    common_contributions[i][si] = 1
+
+            for i in range(j1, j2):
+                if tag == "insert":
+                    helper(src_lines[i], i1)[si] = 1
+                elif tag == "replace":
+                    helper(src_lines[i], i1)[si] = 1
+
+    result: List[Line] = []
+    if merge_mode:
+        for i in range(len(common_lines) + 1):
+            for t, counts in inserted_lines[i].items():
+                lst = [0] * len(rest)
+                for a, b in counts.items():
+                    lst[a] = b
+                result.append(Line(tuple(lst) + (0,), t))
```

Comment:
the zero here represents that it is not part of the common files.  I could have increased the size of `lst` by one but felt like this was slightly less weird.